src/Azure/Orleans.Transactions.AzureStorage/TransactionalState/DistributedTM/AzureTableTransactionalStateStorage.cs

```diff
@@ -0,0 +1,296 @@
+                    for (int i = 0; i < states.Count; i++)
+                    {
+                        states[i].Value.StateJson = null;
+                    }
+
+                    if (logger.IsEnabled(LogLevel.Trace))
+                        logger.LogTrace($"{partition} Loaded v{this.key.CommittedSequenceId} rows={string.Join(",", states.Select(s => s.Key.ToString("x16")))}");
+
+                    return new TransactionalStorageLoadResponse<TState>(this.key.ETag, committedState, this.key.CommittedSequenceId, this.key.Metadata, PrepareRecordsToRecover);
+                }
+            }
+            catch (Exception ex)
+            {
+                this.logger.LogError("Transactional state load failed {Exception}.", ex);
+                throw;
+            }
+        }
+
+
+        public async Task<string> Store(string stateName, string expectedETag, string metadata, List<PendingTransactionState<TState>> statesToPrepare, long? commitUpTo, long? abortAfter)
+        {
+            if (this.key.ETag != expectedETag)
+                throw new ArgumentException(nameof(expectedETag), "Etag does not match");
+
+            // assemble all storage operations into a single batch
+            // these operations must commit in sequence, but not necessarily atomically
+            // so we can split this up if needed
+            var batchOperation = new TableBatchOperation();
+
+            // first, clean up aborted records
+            if (abortAfter.HasValue && states.Count != 0)
+            {
+                while (states.Count > 0 && states[states.Count - 1].Key > abortAfter)
+                {
+                    var entity = states[states.Count - 1].Value;
+                    batchOperation.Delete(entity);
+                    states.RemoveAt(states.Count - 1);
+
+                    if (logger.IsEnabled(LogLevel.Trace))
+                        logger.LogTrace($"{partition}.{states[states.Count - 1].Key:x16} Delete {entity.TransactionId}");
+                }
+            }
+
+            // second, persist non-obsolete prepare records
+            var obsoleteBefore = commitUpTo.HasValue ? commitUpTo.Value : key.CommittedSequenceId;
+            if (statesToPrepare != null)
+                foreach (var s in statesToPrepare)
+                    if (s.SequenceId >= obsoleteBefore)
+                    {
+                        if (FindState(s.SequenceId, out var pos))
+                        {
+                            // overwrite with new pending state
+                            var existing = states[pos].Value;
+                            existing.TransactionId = s.TransactionId;
+                            existing.TransactionTimestamp = s.TimeStamp;
+                            existing.TransactionManager = s.TransactionManager;
+                            existing.SetState(s.State, this.jsonSettings);
+                            batchOperation.Replace(existing);
```

Comment:
Under what conditions would a prepared state change?

Reply:
the prepared states are indexed by sequence id (1,2,3,4...). These sequence ids serve as the row key in the table. When a transaction is aborted, the same sequence is reused by a different transaction.